Fix part of #8472: migrated playthrough-issues-backend-service files from angularjs to angular 8

vinayakvvg opened the pull request:
## Overview
<!--
READ ME FIRST: Please answer *both* questions below and check off every point from the essential checklist!
-->

1. This PR fixes part of #8472 .
2. This PR does the following: Migrated playthrough-issues-backend- api service as well as specs files from angularjs to angular 8; also added it in upgraded services file.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop". 

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

Commit:
migrate playthroug-issues-backend-service files

vinayakvvg commented:
Hi @brianrodri , @bansalnitish which CHANGELOG label do i have to add?
and also I have merged this branch locally with develop branch, why is it still showing resolve conflicts?

kevinlee12 commented:
@vinayakvvg I added the changelog label.

kevinlee12 commented:
@vinayakvvg also mind the merge conflict!

Commit:
Merge conflicts resolved by selecting Incoming changes

vinayakvvg commented:
> @vinayakvvg I added the changelog label.

Thanks @kevinlee12, I have also resolved the merge conflicts.

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -16,26 +16,21 @@
+import { HttpClientTestingModule, HttpTestingController } from
+  '@angular/common/http/testing';
+import { TestBed, fakeAsync, flushMicrotasks } from '@angular/core/testing';
+import { PlaythroughIssuesBackendApiService } from
```
Please add a line break here to distinguish between the angular and Oppia libraries.

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -50,81 +45,66 @@ describe('PlaythroughIssuesBackendApiService', function() {
+      req.flush(backendIssues);
+
+      flushMicrotasks();
+      expect(successHandler).toHaveBeenCalled();
```
Do not replace `toHaveBeenCalledWith` with `toHaveBeenCalled`.

vinayakvvg replied:
But the example given over here at https://github.com/oppia/oppia/wiki/Angular-Migration is doing the same in this file as well topic-viewer-backend-api.service.spec.ts.

If I don't have to replace it then the parameters also should be present inside toHaveBeenCalledWith?

bansalnitish replied:
Yes, the parameters should also be present. Please revert back this to `toHaveBeenCalledWith`.

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -50,81 +45,66 @@ describe('PlaythroughIssuesBackendApiService', function() {
+
+        flushMicrotasks();
+
+        expect(successHandler).toHaveBeenCalled();
```
Ditto here

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -147,68 +127,73 @@ describe('PlaythroughIssuesBackendApiService', function() {
-      expect(successHandler).toHaveBeenCalledWith(
-        PlaythroughObjectFactory.createFromBackendDict(backendPlaythrough));
+      flushMicrotasks();
+      expect(successHandler).toHaveBeenCalled();
```
Ditto here.

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -147,68 +127,73 @@ describe('PlaythroughIssuesBackendApiService', function() {
+        flushMicrotasks();
+
+        expect(successHandler).toHaveBeenCalled();
+        expect(failHandler).not.toHaveBeenCalled();
```
Ditto here

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,100 @@
-          exploration_id: explorationId
-        });
-    };
+import { ServicesConstants } from 'services/services.constants.ts';
```
Move it in between PlayThorughIssueObjectFactory and UrlInterpolationService

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,100 @@
-    };
+@Injectable({
+  providedIn: 'root'
+})
```
Remove the line break below this

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,100 @@
+
+  private cachedIssues = null;
+
+  private getFullIssuesUrl(explorationId: string) {
```
Missing return type of the function

bansalnitish replied:
Ditto for all below functions

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,100 @@
+      });
+  }
+
+  fetchIssues(explorationId: string, explorationVersion: any) {
```
explorationVersion should be `string` right? Please check and remove `any`.

vinayakvvg replied:
@bansalnitish 
I'm confused over here, if I set it as string the spec file starts throwing error as over there fetchIssues is called using a string and a number parameter.

If I change the function declaration to string, number as parameter types to suit the spec file, the get call params inside the function, this -> params: { exp_version: explorationVersion }, starts throwing error, thats why I gave it a any parameter type.

vinayakvvg replied:
@bansalnitish please guide me on this issue

bansalnitish replied:
Hey @vinayakvvg, Thanks for pinging me again -- I think I missed this one.

So, what I did is I looked at the spec file and the line that calls this function. Example this line:
```
playthroughIssuesBackendApiService.fetchIssues('7', 1).then(
  successHandler, failureHandler);
```
which clearly states that first argument is string and second should be number. Could you please change the `explorationVersion` to `number` now?

vinayakvvg replied:
@bansalnitish, hi when I assign number to explorationVersion. It gives the following error while running the typescript_checks. 
--->

The last overload gave the following error.
    Type 'number' is not assignable to type 'string | string[]'.


and  if I assign string to explorationVersion, I get this error.
--->
core/templates/services/playthrough-issues-backend-api.service.spec.ts(68,59): error TS2345: Argument of type '1' is not assignable to parameter of type 'string'.

bansalnitish replied:
I checked this and you're right here. Let's add a comment on top of this something like
```
TODO(#7165): This has been marked any since marking explorationVersion to number throws an error "Type 'number' is not assignable to type 'string | string[]'" and if this is marked to string it throws an error "TS2345: Argument of type '1' is not assignable to parameter of type 'string'".
```

vinayakvvg replied:
@bansalnitish 
Okay, adding the comment.
I'll be raising a PR after this.

Please suggest me some other issue to work on as well.

vinayakvvg replied:
@bansalnitish 
Okay, adding the comment.
I'll be raising a PR after this.

Please suggest me some other issue to work on as well.

bansalnitish replied:
I'd say taken any other service from the issue #8472 itself. There are a lot of services mentioned in the tracker sheet attached there. Thanks!

vinayakvvg replied:
Okay, after selecting a file from tracker, i can just post on this link #8472 to assign that to me right ?

bansalnitish replied:
Assigned!

bansalnitish reviewed (commented):
Thanks @vinayakvvg, the start looks good.

Took the first pass and left comments for you!

Commit:
Resolved review comments.

bansalnitish commented on core/templates/services/UpgradedServices.ts:
```diff
@@ -616,12 +618,18 @@ export class UpgradedServices {
       upgradedServices['SubtopicObjectFactory'],
       upgradedServices['StoryReferenceObjectFactory'],
       upgradedServices['SkillSummaryObjectFactory']);
+    upgradedServices['PlaythroughIssuesBackendApiService'] =
```
Please move it above. Add in the alphabetical order.

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -50,81 +46,69 @@ describe('PlaythroughIssuesBackendApiService', function() {
-        expect(successHandler).toHaveBeenCalledWith(
-          backendIssues.map(
-            PlaythroughIssueObjectFactory.createFromBackendDict));
+        expect(successHandler).toHaveBeenCalled();
```
Please do not change this to toHaveBeenCalled(). The original piece of code was fine. Revert this back to original.

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+  // 'string | string[]'" and if this is marked to string it throws an
+  // error "TS2345: Argument of type '1' is not assignable to parameter
+  // of type 'string'".
+
```
Remove this empty line.

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+  }
+
+  fetchPlaythrough(expId: string, playthroughId: string)
+  : Promise<PlaythroughIssue> {
```
Break this line at the start of opening bracket i.e before `expId:string`.

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+      });
+  }
+
+  resolveIssue(issueToResolve: any, expId: string, expVersion: number)
```
Ditto here

bansalnitish reviewed (commented):
Took another pass, please let me know when you resolve these.

Commit:
Resolved review comments-2.

vinayakvvg commented:
> Took another pass, please let me know when you resolve these.

I have made the changes, please review them again.

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+
+  resolveIssue(
+      issueToResolve: any, expId: string, expVersion: number)
+      : Promise<unknown> {
```
What is this unknown here? Can we replace this with `object`?

bansalnitish commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+  }
+
+  resolveIssue(
+      issueToResolve: any, expId: string, expVersion: number)
```
Why is this marked as `any`?

vinayakvvg replied:
Should I assign object type to "issueToResolve" ?

vinayakvvg replied:
I have fixed these issue, please review them @bansalnitish, Thank you.

Commit:
Resolved review comments - 3

bansalnitish reviewed (approved):
Thanks @vinayakvvg, this looks good to me. Nicely done :)

bansalnitish commented:
Please mind the travis failures.

vinayakvvg commented:
> Thanks @vinayakvvg, this looks good to me. Nicely done :)

Thanks @bansalnitish .
How can i fix the travis issues?

bansalnitish commented:
1. Go through the travis log and see what's failing.
2. Run that individual test suite locally on your machine to see if it failing there too.
3. Then let's discuss what's the issue why it's failing. I'll help you here..

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+        exp_version: expVersion
+      }).toPromise().then(() => {
+      var issueIndex = this.cachedIssues !== null ?
+        this.cachedIssues.findIndex((issue) => {
```
nit: please remove parenthesis around `issue`

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+      expId: string, playthroughId: string): Promise<PlaythroughIssue> {
+    return this.httpClient.get(
+      this.getFullPlaythroughUrl(expId, playthroughId), { observe: 'response' })
+      .toPromise().then((response: any) => {
```
Shouldn't this also have a `TODO(#7165)`?

vinayakvvg replied:
@brianrodri , please elaborate on this.
What I have to do?

brianrodri replied:
Ah, please also leave a TODO comment here (linking to #7165) explaining why we need to give the `response` object the type `any`.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+        var invalidIssueError = new Error(
+          'An issue which was not fetched from the backend has been ' +
+          'resolved');
+        return new Promise((reject) => reject(invalidIssueError));
```
You can just throw an error, Javascript will transform it into a "reject" for you (can confirm with unit tests)

vinayakvvg replied:
> 1. Go through the travis log and see what's failing.
> 2. Run that individual test suite locally on your machine to see if it failing there too.
> 3. Then let's discuss what's the issue why it's failing. I'll help you here..

Okay @bansalnitish  will try the above mentioned points.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+  // of type 'string'".
+  fetchIssues(explorationId: string, explorationVersion: any): Promise<string> {
+    if (this.cachedIssues !== null) {
+      return new Promise((resolve) => resolve(this.cachedIssues));
```
Please prefer `$q` to `Promise`. Also, this can be simplified to:

```typescript
if (this.cachedIssues !== null) {
  return $q.resolve(this.cachedIssues);
}
```

vinayakvvg replied:
I referenced this example ->
https://github.com/oppia/oppia/pull/8427/files,
that's why changed it into a Promise, what is the correct approach?

brianrodri replied:
We should import the `$q` service, and use that to create Promise instances.

vinayakvvg replied:
Hi @brianrodri ,
Why are we still using $q if it was a part of AngularJS,
this is a migration to Angular 8 PR, hence we should be using Promises.

Can you explain on this, and if I am wrong then can you provide any reference to read about how $q is used in Angular 8.

bansalnitish replied:
$q is something that works with AngularJS. For Angular we are using new Promise.

brianrodri replied:
Got it, in that case please use `Promise.resolve(value)` instead ([source](https://codecraft.tv/courses/angular/es6-typescript/promises/#_immediate_resolution_or_rejection)).

brianrodri replied:
This still isn't done? Please change to:

```typescript
if (this.cachedIssues !== null) {
  return Promise.resolve(this.cachedIssues);
}
```

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+      return this.httpClient.get(
+        this.getFullIssuesUrl(explorationId), {
+          params: { exp_version: explorationVersion }, observe: 'response'
+        }).toPromise().then((response: any) => {
```
Shouldn't this also have a `TODO(#7165)`?

vinayakvvg replied:
@brianrodri , I am confused here as well.

brianrodri replied:
Ditto, please leave a comment explaining why `response` needs to have the type `any`.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+      });
+  }
+
+  private getFullPlaythroughUrl(expId: string, playthroughId: string): string {
```
Please change `expId` to `explorationId` or vise versa, we should keep the name consistent throughout the entire file.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+export class PlaythroughIssuesBackendApiService {
+  constructor(
+    private httpClient: HttpClient,
+    private urlInterpolationService: UrlInterpolationService,
```
Please sort these.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+  }
+
+  fetchPlaythrough(
+      expId: string, playthroughId: string): Promise<PlaythroughIssue> {
```
Ditto. Change all `expId` to `explorationId`, or vice versa, but do not leave the names inconsistent.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,106 @@
+    return this.httpClient.get(
+      this.getFullPlaythroughUrl(expId, playthroughId), { observe: 'response' })
+      .toPromise().then((response: any) => {
+        var playthroughBackendDict = response.body;
```
Please use `let` instead of `var`, here and elsewhere.

brianrodri reviewed (changes requested):
Took a pass, thanks @vinayakvvg!

vinayakvvg commented:
> 1. Go through the travis log and see what's failing.
> 2. Run that individual test suite locally on your machine to see if it failing there too.
> 3. Then let's discuss what's the issue why it's failing. I'll help you here..

Okay @bansalnitish will try the above mentioned points.

Commit:
Resolved review comments - 3

Commit:
Merge branch 'develop' into migrate-playthrough-issues-backend

vinayakvvg commented:
> Took a pass, thanks @vinayakvvg!

I have made the changes, PTAL @brianrodri

vinayakvvg commented:
> > 1. Go through the travis log and see what's failing.
> > 2. Run that individual test suite locally on your machine to see if it failing there too.
> > 3. Then let's discuss what's the issue why it's failing. I'll help you here..
> 
> Okay @bansalnitish will try the above mentioned points.

Hi @bansalnitish,

On executing the following command I am getting this error 
-->> 
Command: python -m scripts.run_e2e_tests --skip-install --skip-build --suite="classroomPageFileUploadFeatures"

Error: -->>

1) Classroom page functionality should add a new published topic to the Math classroom
  - Failed: Could not set role successfully
  Wait timed out after 10004ms
  - Failed: Create Topic button takes too long to be clickable
  Wait timed out after 10005ms
  - Expected $.length = 1 to equal 0.
  Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/topics_and_skills_dashboard - Failed to load resource: the server responded with a status of 401 (Unauthorized)', timestamp: 1585711001152, type: '' }) to equal undefined.

I guess this is not a part of this PR, regardless please guide me how to resolve this.

bansalnitish commented:
I'll try to debug this for you and get back to you. Thanks!

bansalnitish commented:
Hi @vinayakvvg, Probably your changes require interceptor as well. I've added interceptor in all the modules in #8965. Once that goes in, please merge the latest develop and hopefully your tests will pass :).

Alternatively, if you don't want to wait for that to go in -- you can add the interceptor that is concerned with this particular service.

bansalnitish reviewed (approved):
LGTM (the code changes)!

Commit:
Merge branch 'develop' into migrate-playthrough-issues-backend

vinayakvvg commented:
> Hi @vinayakvvg, Probably your changes require interceptor as well. I've added interceptor in all the modules in #8965. Once that goes in, please merge the latest develop and hopefully your tests will pass :).
> 
> Alternatively, if you don't want to wait for that to go in -- you can add the interceptor that is concerned with this particular service.

Thanks @bansalnitish , I guess #8965 is merged I will take a pull from develop then push my branch,

I am getting this error from past one day, when I execute this statement -->

python -m scripts.pre_commit_linter
Error -->> 
/System/Library/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python: No module named scripts.pre_commit_linter

has something changed?

Thanks!

Commit:
Merge branch 'develop' into migrate-playthrough-issues-backend

vinayakvvg commented:
> > Hi @vinayakvvg, Probably your changes require interceptor as well. I've added interceptor in all the modules in #8965. Once that goes in, please merge the latest develop and hopefully your tests will pass :).
> > Alternatively, if you don't want to wait for that to go in -- you can add the interceptor that is concerned with this particular service.
> 
> Thanks @bansalnitish , I guess #8965 is merged I will take a pull from develop then push my branch,
> 
> I am getting this error from past one day, when I execute this statement -->
> 
> python -m scripts.pre_commit_linter
> Error -->>
> /System/Library/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python: No module named scripts.pre_commit_linter
> 
> has something changed?
> 
> Thanks!

Hi @bansalnitish 
I merged the changes from develop branch still there was build faiilures due to some permission access.

Getting this as error message -->

git@github.com: Permission denied (publickey).

fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

How can I resolve these issues?

vojtechjelinek commented:
@vinayakvvg What command gives you this error?

vinayakvvg commented:
> @vinayakvvg What command gives you this error?

Hi @vojtechjelinek, 

These are the error logs from the circleci build failures, all three of them have the same errors.
How to resolve these?

vojtechjelinek commented:
@vinayakvvg Ah, okay, sorry. I have restarted the tests these are probably just some flaky runs.

vinayakvvg commented:
> @vinayakvvg Ah, okay, sorry. I have restarted the tests these are probably just some flaky runs.

@vojtechjelinek, @bansalnitish 
 
**1. some backend codecov is failing, in which i have not made any changes
2. travis CI build - got stuck waiting for some action, i think it is stuck at #25616.20
3. Travis build #25616.10 - on running the following command I got these errors
cmd: python -m scripts.run_e2e_tests --suite=explorationImprovementsTab --prod_env**

errors:
i) Playthrough Improvements Early Quit playthroughs resolving the task disappears from the improvements tab
  - Failed: element not visible
    (Session info: chrome=80.0.3987.149)
    (Driver info: chromedriver=2.41.578706 (5f725d1b4f0a4acbf5259df887244095596231db),platform=Mac OS X 10.15.3 x86_64)
  - Failed: Library Page does not have any explorations
  Wait timed out after 10004ms

ii) Playthrough Improvements Early Quit playthroughs viewing the task hides the mark as resolve button from guests
  - Failed: element not visible
    (Session info: chrome=80.0.3987.149)
    (Driver info: chromedriver=2.41.578706 (5f725d1b4f0a4acbf5259df887244095596231db),platform=Mac OS X 10.15.3 x86_64)
  - Failed: Library Page does not have any explorations
  Wait timed out after 10005ms
  - Failed: Improvements tab is not clickable
  Wait timed out after 10001ms

**4. Travis build #25616.11 - on running the following command I got these errors
cmd: python -m scripts.run_e2e_tests --suite=explorationStatisticsTab --prod_env**

errors:
i) Issues visualization records early quit issue.
  - Failed: No element found using locator: By(css selector, .protractor-test-issue0)
  - Expected $.length = 1 to equal 0.
  Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js 126:302 TypeError: Cannot read property 'interpolateUrl' of undefined
      at e.getFullIssuesUrl (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:1:935250)
      at e.fetchIssues (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:1:935818)
      at Object.getIssues (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:1:1559761)
      at controller.$onInit (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:12:255330)
      at http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:72:271
      at r (http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:8:170)
      at p (http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:72:136)
      at http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:83:390
      at http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:140:44
      at m.$digest (http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:151:185)', timestamp: 1586105636450, type: '' }) to equal undefined.

ii) Issues visualization records multiple incorrect issue.
  - Failed: No element found using locator: By(css selector, .protractor-test-issue0)
  - Expected $.length = 1 to equal 0.
  Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js 126:302 TypeError: Cannot read property 'interpolateUrl' of undefined
      at e.getFullIssuesUrl (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:1:935250)
      at e.fetchIssues (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:1:935818)
      at Object.getIssues (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:1:1559761)
      at controller.$onInit (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:12:255330)
      at http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:72:271
      at r (http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:8:170)
      at p (http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:72:136)
      at http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:83:390
      at http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:140:44
      at m.$digest (http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:151:185)', timestamp: 1586105660226, type: '' }) to equal undefined.

iii) Issues visualization records cyclic transitions issue.
  - Failed: No element found using locator: By(css selector, .protractor-test-issue0)
  - Expected $.length = 1 to equal 0.
  Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js 126:302 TypeError: Cannot read property 'interpolateUrl' of undefined
      at e.getFullIssuesUrl (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:1:935250)
      at e.fetchIssues (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:1:935818)
      at Object.getIssues (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:1:1559761)
      at controller.$onInit (http://localhost:9001/build/webpack_bundles/exploration_editor.44dfa79db848bd2a751e.bundle.js:12:255330)
      at http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:72:271
      at r (http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:8:170)
      at p (http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:72:136)
      at http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:83:390
      at http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:140:44
      at m.$digest (http://localhost:9001/third_party/static/angularjs-1.7.9/angular.min.js:151:185)', timestamp: 1586105708601, type: '' }) to equal undefined.

iv) Statistics tab checks statistics tab for an exploration
  - Failed: element not visible
    (Session info: chrome=80.0.3987.149)
    (Driver info: chromedriver=2.41.578706 (5f725d1b4f0a4acbf5259df887244095596231db),platform=Mac OS X 10.15.3 x86_64

...............................
Please help me in resolving these issues??

vojtechjelinek commented:
@vinayakvvg As I said, the problem is that these tests are flaky (sometimes fail without reasons related to the changes you made). I have restarted and most of them are passing now, there is some slight problem with Travis but I have cleared the cahce and it should be working now. I will ping you if I need any assistance from you.

vojtechjelinek commented:
@brianrodri PTAL

brianrodri commented on core/templates/services/UpgradedServices.ts:
```diff
@@ -606,15 +608,21 @@ export class UpgradedServices {
     upgradedServices['PlayerPositionService'] = new PlayerPositionService(
       upgradedServices['ContextService'],
       upgradedServices['PlayerTranscriptService']);
+    upgradedServices['PlaythroughIssuesBackendApiService'] =
```
Please confirm whether this is in the correct `Topological level: #` group.

vinayakvvg replied:
Earlier there were instructions in the `UpgradedServices.ts` file on how to add the new services based on their dependencies. I followed those. 
In the instructions the requirement was that all the dependecnies of the current service should  already be present in the above Topological level.

Correct me if I am wrong.

brianrodri replied:
Nope that sounds good to me, thanks for confirming!

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -16,26 +16,22 @@
+describe('PlaythroughIssuesBackendApiService', () => {
+  let playthroughIssuesBackendApiService:
+    PlaythroughIssuesBackendApiService = null;
+  let httpTestingController: HttpTestingController;
```
Please set either everything to `null`, or nothing to `null`. Do not mix conventions.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -16,26 +16,22 @@
+  let httpTestingController: HttpTestingController;
+
+  let playthroughIssueObjectFactory: PlaythroughIssueObjectFactory = null;
+  let backendIssues = [{
```
Please group logically-similar objects together, and split each group by newlines.

Here, the logically-similar objects would be the factories/services vs. the test-data.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -50,81 +46,70 @@ describe('PlaythroughIssuesBackendApiService', function() {
-        '/issuesdatahandler/7?exp_version=1'
-      ).respond(backendIssues);
+  describe('.fetch', () => {
+    it('returns the issues data provided by the backend', fakeAsync(() => {
```
Please update these spec descriptions to follow our testing convention (`it` descriptions always start with `'should '`):

```typescript
it('should return the thing', () => {
});
```

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.spec.ts:
```diff
@@ -50,81 +46,70 @@ describe('PlaythroughIssuesBackendApiService', function() {
 
         // Try to fetch another issue
-        PlaythroughIssuesBackendApiService.fetchIssues('8', 1).then(
+        playthroughIssuesBackendApiService.fetchIssues('8', 1).then(
```
Shouldn't this need another call to `flushMicrotasks`?

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,110 @@
-require('services/services.constants.ajs.ts');
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+import { HttpClient, HttpResponse } from '@angular/common/http';
```
Please sort these lines.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,110 @@
+  constructor(
+    private httpClient: HttpClient,
+    private playthroughIssueObjectFactory: PlaythroughIssueObjectFactory,
+    private urlInterpolationService: UrlInterpolationService) { }
```
nit: remove space between braces.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,110 @@
+    } else {
+      return this.httpClient.get(
+        this.getFullIssuesUrl(explorationId), {
+          params: { exp_version: explorationVersion }, observe: 'response'
```
Please move `observe` onto its own line.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,110 @@
+        this.getFullIssuesUrl(explorationId), {
+          params: { exp_version: explorationVersion }, observe: 'response'
+        }).toPromise().then(
+        (response: HttpResponse<Array<PlaythroughIssue>>) => {
```
Please prefer `Type[]` to `Array<Type>`.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,110 @@
+  fetchPlaythrough(
+      explorationId: string, playthroughId: string): Promise<PlaythroughIssue> {
+    return this.httpClient.get(
+      this.getFullPlaythroughUrl(
```
Please prefer keeping the opening of config-objects on the line of the URL, for example, please change this to:
```typescript
return this.httpClient.get(
  this.getFullPlaythroughUrl(explorationId, playthroughId), {
    observe: 'response'
  })).toPromise().then(...);
```

brianrodri reviewed (changes requested):
Took my first pass, PTAL.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,110 @@
+
+  resolveIssue(
+      issueToResolve: PlaythroughIssue, explorationId: string,
+      expVersion: number): Promise<object> {
```
Please add `TODO(#7165):` explaining why this is returning `Promise<object>`.

vinayakvvg replied:
Hi @brianrodri i am not sure how I can find this out, please guide me. Thanks.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,110 @@
+
+  resolveIssue(
+      issueToResolve: PlaythroughIssue, explorationId: string,
+      expVersion: number): Promise<object> {
```
Please prefer using `any` to `object`.

vinayakvvg replied:
Earlier I was using `Promise<unknown>` and @bansalnitish during earlier review comments asked me to change this to `Promise<object>`.

@brianrodri, @bansalnitish 
Please explain and clarify the correct approach, thanks.

brianrodri replied:
I'd like to see us use types that are more common in the codebase, and `any` is more common than `object`. So please use `Promise<any>` instead.

vinayakvvg commented:
> @vinayakvvg As I said, the problem is that these tests are flaky (sometimes fail without reasons related to the changes you made). I have restarted and most of them are passing now, there is some slight problem with Travis but I have cleared the cahce and it should be working now. I will ping you if I need any assistance from you.

Thanks @vojtechjelinek

vinayakvvg commented:
> @vinayakvvg What command gives you this error?
Hi @vojtechjelinek, one more issue
 
After merging from develop branch locally, I am getting this error from a couple of days, when I execute this statement -->

cmd: `python -m scripts.pre_commit_linter`
Error:
`/System/Library/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python: No module named scripts.pre_commit_linter`

has something changed?

Commit:
Resolved review comments - 4

bansalnitish commented:
Yes, please use `python -m scripts.linters.pre_commit_linter`.

There are some more flags that can be used with this command now. Refer to https://github.com/oppia/oppia/blob/develop/scripts/linters/pre_commit_linter.py file.

bansalnitish commented:
Also, the tests pass on this PR now :). Could you address Brian's comment please and ping him when you're ready for another pass?

Commit:
Resolved review comments - 4

vinayakvvg commented:
> Yes, please use `python -m scripts.linters.pre_commit_linter`.
> 
> There are some more flags that can be used with this command now. Refer to https://github.com/oppia/oppia/blob/develop/scripts/linters/pre_commit_linter.py file.

Thanks @bansalnitish, I found it after looking the file structures.

Please update this on this link as well https://github.com/oppia/oppia/wiki/Angular-Migration#overview
As anyone new coming to this page would find it easier to run this command.

vinayakvvg commented:
> Also, the tests pass on this PR now :). Could you address Brian's comment please and ping him when you're ready for another pass?

Sure, @brianrodri I have addressed your comments PTAL.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -86,18 +87,18 @@ export class PlaythroughIssuesBackendApiService {
-      });
+      this.getFullPlaythroughUrl(explorationId, playthroughId), {
+        observe: 'response'
+      }).toPromise().then((response: HttpResponse<PlaythroughIssue>) => {
```
This seems wrong to me? The `HttpResponse` would have a backend dict, not a fully constructed `PlaythroughIssue` instance. Please update the type of these `HttpResponse` values to `any`, and leave a `TODO(#7165)` for giving them a proper type.

vinayakvvg replied:
Hi @brianrodri I have assigned the response as `PlaythroughIssue` based on the same function call from `spec.ts` file.
The response that we get over there from calling `createFromBackendDict` function is an instance of `PlaythroughIssue` (which we get form `PlaythroughIssueObjectFactory.ts`).

I took this approach in determining the response type of a particular http response, if I am wrong then help me in determining the correct approach of finding a http response type.

PTAL, thanks.

brianrodri replied:
Right, but this response hasn't been passed along to `createFromBackendDict` yet. Therefore you are in the "middle" of construction, where you are still holding onto a bare backend dict which is `then` converted into a `PlaythroughIssue`.

In other words, here's the issue illustrated by breaking up the promise chain.

```typescript
// TODO(#7165): Change `any` to a type describing the dict.
let httpResponse: HttpResponse<any> = this.httpClient.get(
  url, options);

let playthroughIssuePromise: Promise<PlaythroughIssue> =
  httpResponse.toPromise().then(httpResponse => {
    let backendDict = httpResponse.body; // **NOT** a PlaythroughIssue!
    return PlaythroughIssueObjectFactory.createFromBackendDict(
      backendDict);
  });

return playthroughIssuePromise;
```

If the `spec` says otherwise, then it will need to be updated so it uses the correct types.

vinayakvvg replied:
Understood, thanks, I'll update the PR.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,111 @@
+          params: { exp_version: explorationVersion },
+          observe: 'response'
+        }).toPromise().then(
+        (response: HttpResponse<PlaythroughIssue[]>) => {
```
Ditto, this should be an `HttpResponse<any[]>`, and have a `TODO(#7165)` to update them to use a valid type.

vinayakvvg replied:
Same doubt, over here as well.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,111 @@
+        exp_issue_dict: issueToResolve.toBackendDict(),
+        exp_version: expVersion
+      }).toPromise().then(() => {
+      let issueIndex = this.cachedIssues !== null ?
```
This is a bit hard to read, please convert this into a regular if statement instead.

Commit:
Review comments addressed - 5

vinayakvvg commented:
@brianrodri, I have updated the PR PTAL, thanks.

Commit:
Review comments addressed - 5

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -89,7 +89,7 @@ export class PlaythroughIssuesBackendApiService {
       this.getFullPlaythroughUrl(explorationId, playthroughId), {
         observe: 'response'
-      }).toPromise().then((response: HttpResponse<PlaythroughIssue>) => {
+      }).toPromise().then((response: HttpResponse<any>) => {
```
Please also add the TODO to change this to a specific type (here and elsewhere we use `any`).

vinayakvvg replied:
For Httpresponse I have made these changes

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -104,10 +104,14 @@ export class PlaythroughIssuesBackendApiService {
+      } else {
+        issueIndex = -1;
+      }
       if (issueIndex === -1) {
```
OK, this code is much clearer to me now, thanks! That said, I think it can be made even more clear - please change the `then` block to:

```typescript
if (this.cachedIssues !== null) {
  let issueIndex = this.cachedIssues.findIndex(
    issue => angular.equals(issue, issueToResolve));
  if (issueIndex !== -1) {
    this.cachedIssues.splice(issueIndex, 1);
    return;
  }
}
throw Error(
  'An issue which was not fetched from the backend has been ' +
  'resolved');
```

vinayakvvg replied:
This is throwing frontend_tests errors, from spec.ts file of spy success and failure.

brianrodri replied:
OK, I'll just do it in a separate PR myself then. Feel free to disregard this comment.

vinayakvvg replied:
This one as well, please add me to that PR.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,117 @@
-    };
+  private cachedIssues = null;
+
+  private getFullIssuesUrl(explorationId: string): string {
```
Please move all private functions to the end of the class, and public functions to the beginning (but after the constructor).

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,117 @@
+
+  resolveIssue(
+      issueToResolve: PlaythroughIssue, explorationId: string,
+      expVersion: number): Promise<any> {
```
According to the success branch we aren't actually returning anything, right? Seems to me, then, that this should be changed to have a return type of `Promise<void>`.

vinayakvvg replied:
If i change it to void it throws this error
`error TS2322: Type 'Promise<void | Observable<never>>' is not assignable to type 'Promise<void>'.`

 ` Type 'void | Observable<never>' is not assignable to type 'void'.`

  `  Type 'Observable<never>' is not assignable to type 'void'.`

brianrodri replied:
OK, I'll look into it myself in a separate PR. Feel free to disregard this comment.

vinayakvvg replied:
Please add me onto that PR, I would like to know how this is solved, thanks

brianrodri reviewed (approved):
Looks great, thanks again @vinayakvvg! This is the final review AFAICT, please resolve these and I think it'll be ready to merge :)

Commit:
Review comments addressed - 6

vinayakvvg commented:
Hi @brianrodri, I have made the changes, PTAL.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,117 @@
+
+  resolveIssue(
+      issueToResolve: PlaythroughIssue, explorationId: string,
+      expVersion: number): Promise<any> {
```
Please rename to `explorationVersion` so it matches the convention of `explorationId`, here and elsewhere.

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,117 @@
-          playthrough_id: playthroughId
-        });
-    };
+  private cachedIssues = null;
```
Please move the instance variables to the very beginning of the class (before the constructor).

brianrodri reviewed (approved):
LGTM, I'll handle the suggested changes myself once this PR is merged. Thanks!

kevinlee12 commented:
@brianrodri are you implying that you will be making a follow up PR?

brianrodri commented:
> @brianrodri are you implying that you will be making a follow up PR?

Yup that's right

kevinlee12 commented:
Okay, I’ll mark this as lgtm then.

brianrodri reviewed (changes requested):
Ah, please resolve my last two comments first though.

Commit:
Review comments addressed - 7

vinayakvvg commented:
> Ah, please resolve my last two comments first though.

Hi @brianrodri, I have made the changes, PTAL

brianrodri commented on core/templates/services/playthrough-issues-backend-api.service.ts:
```diff
@@ -15,87 +15,117 @@
+  // 'string | string[]'" and if this is marked to string it throws an
+  // error "TS2345: Argument of type '1' is not assignable to parameter
+  // of type 'string'".
+  fetchIssues(explorationId: string, explorationVersion: any): Promise<string> {
```
Wasn't this `Promise<PlaythroughIssues[]>` before? Seems you changed it to `Promise<string>` by accident? Please change it back.

vinayakvvg replied:
I think it was like this from the beginning, still I have updated it. :)

Commit:
Review comments addressed - 7

vinayakvvg commented:
@brianrodri, what will be the next step?

brianrodri commented:
@vinayakvvg nothing, the PR looks good to me :)

vinayakvvg commented:
Thanks @bansalnitish, @brianrodri, @vojtechjelinek for reviewing and guiding through my first PR.

bansalnitish commented:
You're welcome @vinayakvvg :D.
